### ISSUE TITLE
Support AWS Lambda Node 8.10's async/await during invoke local

### ISSUE THREAD

**Issue by lewisf** [serverless/serverless#4911]
# This is a Bug Report

#### What went wrong?

Ran `invoke local` on a function written using async/await and the node8.10 runtime. It didn't run successfully since serverless still seems to expect a `callback` function.

#### What did you expect should have happened?

It takes the Promise resulting from the async lambda function and waits for it to succeed.

#### What was the config you used?

```
provider:
  name: aws
  runtime: nodejs8.10
```

#### What stacktrace or error message from your provider did you see?

`invoke local` just exits early, instead of waiting for the async stuff to finish, leading to no response.

#### Other details

* ***Serverless Framework Version you're using***: 1.26.1
* ***Operating System***: Mac OS 10.13.4
* ***Stack Trace***: N/A
* ***Provider Error messages***: N/A

**Comment by HyperBrain** [comment]
Hi @lewisf ,
thanks for raising the issue. Regarding a possible implementation, I think returning a promise in lambda functions is not handled well in general.
You're allowed to return promises with all available Node runtimes.
A proper fix would be to store the result of the executed handler function and test if it is a promise or not. If not, await the callback, if it is, `then` the promise and wait for the resolution.
This will fix the issue for all available Node versions (as await/async also returns a promise under the hood).

**Comment by XertroV** [comment]
Stumbled upon this issue - I'm using sls (1.26.1) with async/await and local invocation seems to work fine. (node v8.8.1 installed locally)

**Comment by dboune** [comment]
async/await functions themselves work fine, and really invoking locally does complete the task as expected (it is not exiting early), but any result returned from the function is not printed to the console as it would be if the callback were executed, which could create a perception of exiting early.

Given a function such as:

```js
const lambdaFunction = async () => {
  try {
    // Do something with a promise
    return { status: 'success' }
  } catch (err) {
    // Handle or return error
    return err
  }
}
```

We would then expect to see either the return value or the error printed to the console as it would be using the callback.

This would only be the expected behavior if `runtime: nodejs8.10` (or newer).

Things get a bit more complex when considering plugins like serverless-offline, but I'm not myself aware of how they handle invocation, or if they depend on some handling internal to serverless to receive the function result.

In the end, I believe a description of the expected handling goes something like:

> If the callback is executed under any circumstances, print the error or data from the callback, otherwise if the runtime is nodejs8.10 or above, print any available return value from the lambda function.

I imagine this works the same even without async/promises using the 8.10 runtime, though I haven't tested it, which would also mean the 8.10 runtime wouldn't likely bother to wait around to present an error if the callback isn't executed.

**Comment by HyperBrain** [comment]
@XertroV @dboune Can you verify, that PR #4912 introduces the expected behavior for `sls invoke local`?

**Comment by HyperBrain** [comment]
@daniel-cottone I assume that using handlers that return promises instead of using the callback will need a feature implementation in `serverless-offline`?

**Comment by daniel-cottone** [comment]
@HyperBrain Yes, we have a PR open that should solve this: https://github.com/dherault/serverless-offline/pull/398

**Comment by dboune** [comment]
@HyperBrain - The code did function as expected when returning a value or throwing an error from within async/promise, without using the callback.

Noting that if I return an error object from the function directly, nothing is output, but throw does cause output of the error.

If I instead use the callback from inside a promise, which should be valid, I received the following error which does not occur with 1.26.1:
```json
{
    "errorMessage": "Cannot read property 'headers' of undefined",
    "errorType": "TypeError"
}
```

The above error goes away if I also return something after the callback, and I then receive two outputs.

I believe that if the callback is fired, any other return value should be ignored, especially for the sake of backward compatibility. It would make sense for the callback to take precedence.

**Comment by HyperBrain** [comment]
@dboune Thanks for testing. @lewisf Can you have a look at his comment and try to solve these issues?

**Comment by entrptaher** [comment]
Bumping this issue, this error is bothersome. Please fix, thanks.

**Comment by lewisf** [comment]
Finally got around to addressing. Updated #4912

**Comment by tomasdev** [comment]
OMG

Please merge <3
